### DISCUSSION TITLE
Heltec CubeCell HTCC-AB02S with 'failed, code -5' using LoRa

### DISCUSSION THREAD

**Discussion by javrolijk** [timmbogner/Farm-Data-Relay-System#178] [Q&A]
Hi all,

I'm trying to make the Heltec Cubecell HTCC-AB02S send random sensor data with LoRa to an MQTT gateway. While I can see the data being received in MQTT, it gives a **'failed, code -5'** error after a while and stops sending data. Here's the latest Serial Monitor run. Any clues as to what is going on here? Could it be a faulty board? Below I've also included the 'fdrs_node_config.h' as a reference.

21:15:42.490 -> FDRS User Node initializing...
21:15:42.490 ->  Reading ID 2
21:15:42.490 ->  Gateway: 1
21:15:42.614 -> RadioLib initialization successful!
21:15:42.614 -> LoRa Initialized. Frequency: 915.00  Bandwidth: 125.00  SF: 10  CR: 5  SyncWord: 18  Tx Power: 17dBm
21:15:42.646 -> LoRa node address is 7839 (hex).
21:15:42.646 -> Id: 2 - Type: 3 - Data loaded: 74.00
21:15:42.646 -> Id: 2 - Type: 1 - Data loaded: 21.00
21:15:42.646 -> Sending FDRS Packet!
21:15:42.691 -> Transmitting LoRa message of size 20 bytes with CRC 0xb147 to gateway 0xee01. Retries remaining: 2
21:15:43.417 -> Incoming LoRa. Size: 11 Bytes, RSSI: -23.00dBm, SNR: 7.75dB, PacketCRC: 0xdfde, Total LoRa received: 0
21:15:43.417 -> ACK Received - CRC Match
21:15:43.417 ->  Deep sleeping.
21:15:43.417 ->  Delaying.
21:16:13.883 -> Id: 2 - Type: 3 - Data loaded: 80.00
21:16:13.883 -> Id: 2 - Type: 1 - Data loaded: 21.00
21:16:13.883 -> Sending FDRS Packet!
21:16:13.883 -> Transmitting LoRa message of size 20 bytes with CRC 0x3ef6 to gateway 0xee01. Retries remaining: 2, Ack Ok 100.00%
21:16:14.268 ->  failed, code -5

**The 'fdrs_node_config.h' content:**
`//  FARM DATA RELAY SYSTEM
//
//  Sensor Configuration

#include <fdrs_globals.h>

#define READING_ID    2   //Unique ID for this sensor
#define GTWY_MAC      0x01 //Address of the nearest gateway

//#define USE_ESPNOW
#define USE_LORA
#define DEEP_SLEEP
//#define POWER_CTRL    14
#define FDRS_DEBUG

// LoRa Configuration
#define RADIOLIB_MODULE SX1262
#define LORA_SS    RADIO_NSS
#define LORA_RST   RADIO_RESET
#define LORA_DIO   RADIO_DIO_1
#define LORA_BUSY  RADIO_BUSY
#define USE_SX126X

#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK        // Request LoRa acknowledgment.

//#define CUSTOM_SPI
//#define LORA_SPI_SCK 5
//#define LORA_SPI_MISO 19
//#define LORA_SPI_MOSI 27`

**Comment by timmbogner** [comment]
Hi Jairo,
First try disabling ```DEEP_SLEEP```. The deep-sleeping routines are specific to ESP32 and 8266. If you need lower power consumption, we could look into adding chip-specific code for doing that on CubeCell.
@sbobowski

**Reply by javrolijk** [reply]
Hi Timm,
Thanks for the quick response.
Here's a clue: I commented out '#define LORA_ACK', and it's currently sending data to the MQTT gateway for a couple of hours now without the 'failed, code -5'. Also, if the low-power feature isn't working for the Cubcell, I would be very much interested in the code for this board.

Any other thoughts are much appreciated.

**Reply by timmbogner** [reply]
I'll keep looking into it. For reference, here are the [error code explanations.](https://jgromes.github.io/RadioLib/group__status__codes.html)

**Reply by javrolijk** [reply]
@timmbogner quick question. Did you, by any chance had a chance to look into the code -5 error? I've tested out running FDRS with '#define LORA_ACK' commented out and all work Ok. However, no Ack is received.

**Reply by timmbogner** [reply]
Hey, sorry to leave you hanging there. I forgot this was ongoing. 

Does the problem happen randomly, or does it occur after a certain amount of cycles?  Have you had a chance to test on multiple CubeCell boards to rule out a hardware issue? 

I know that the developer of RadioLib faced some challenges supporting CubeCell, so I'm open to it being a larger issue. However, there is probably greater chance I've overlooked something.
If you haven't already, see if you can get these examples running on  your CubeCells:

https://github.com/timmbogner/FDRS_LoRa_Test

**Reply by javrolijk** [reply]
Hi Timm,

Below is a copy of the Serial Monitor output up to the 'failed, code -5' hard stop. It seems like it goes for 4 cycles before giving the error code. I've tried increasing the time in '#define GLOBAL_ACK_TIMEOUT 600' in 'fdrs_node_lora.h' but no joy.
Unfortunately, I have only one Cubecell to test currently. But I intend to acquire just the regular DIYmalls LoRa ESP32 OLED board if the Cubecell continues with the error. I like that the Cubecell already integrates a GPS, though.
Any help is much appreciated.

23:30:03.714 -> FDRS User Node initializing...
23:30:03.714 ->  Reading ID 1
23:30:03.714 ->  Gateway: 2
23:30:03.808 -> RadioLib initialization successful!
23:30:03.808 -> LoRa Initialized. Frequency: 915.00  Bandwidth: 125.00  SF: 10  CR: 5  SyncWord: 18  Tx Power: 20dBm
23:30:03.846 -> LoRa node address is 93a6 (hex).
23:30:03.884 -> Found SQM (TSL) sensor
23:30:03.884 -> Integration: 200.00 ms
23:30:03.884 -> Gain:        1.00x
23:30:03.884 -> Id: 1 - Type: 1 - Data loaded: 29.91
23:30:03.884 -> Id: 1 - Type: 3 - Data loaded: 68.00
23:30:03.884 -> Id: 1 - Type: 4 - Data loaded: 1013.44
23:30:03.884 -> Id: 1 - Type: 23 - Data loaded: -1.58
23:30:04.166 -> Odd, full less than ir!  Rechecking...
23:30:04.390 -> Bumping gain up
23:30:04.390 -> Integration: 200.00 ms
23:30:04.390 -> Gain:        25.00x
23:30:04.944 -> Id: 1 - Type: 34 - Data loaded: 8.16
23:30:05.268 -> Sending FDRS Packet!
23:30:05.268 -> Transmitting LoRa message of size 41 bytes with CRC 0x9197 to gateway 0xee02. Retries remaining: 2
23:30:06.166 -> Incoming LoRa. Size: 11 Bytes, RSSI: -37.00dBm, SNR: 8.25dB, PacketCRC: 0x9a75, Total LoRa received: 0
23:30:06.166 -> ACK Received - CRC Match
23:31:07.062 -> Id: 1 - Type: 1 - Data loaded: 29.92
23:31:07.062 -> Id: 1 - Type: 3 - Data loaded: 68.00
23:31:07.062 -> Id: 1 - Type: 4 - Data loaded: 1013.39
23:31:07.062 -> Id: 1 - Type: 23 - Data loaded: -1.17
23:31:07.327 -> Id: 1 - Type: 34 - Data loaded: 8.16
23:31:07.637 -> Sending FDRS Packet!
23:31:07.637 -> Transmitting LoRa message of size 41 bytes with CRC 0xfb6d to gateway 0xee02. Retries remaining: 2, Ack Ok 100.00%
23:31:08.495 -> Incoming LoRa. Size: 11 Bytes, RSSI: -37.00dBm, SNR: 7.75dB, PacketCRC: 0x9a75, Total LoRa received: 1, CRC Ok Pct 100.00%
23:31:08.541 -> ACK Received - CRC Match
23:32:09.430 -> Id: 1 - Type: 1 - Data loaded: 29.93
23:32:09.430 -> Id: 1 - Type: 3 - Data loaded: 68.00
23:32:09.430 -> Id: 1 - Type: 4 - Data loaded: 1013.40
23:32:09.430 -> Id: 1 - Type: 23 - Data loaded: -1.25
23:32:09.648 -> Id: 1 - Type: 34 - Data loaded: 8.17
23:32:10.000 -> Sending FDRS Packet!
23:32:10.001 -> Transmitting LoRa message of size 41 bytes with CRC 0x321a to gateway 0xee02. Retries remaining: 2, Ack Ok 100.00%
23:32:10.862 -> Incoming LoRa. Size: 11 Bytes, RSSI: -37.00dBm, SNR: 7.75dB, PacketCRC: 0x9a75, Total LoRa received: 2, CRC Ok Pct 100.00%
23:32:10.862 -> ACK Received - CRC Match
23:33:11.785 -> Id: 1 - Type: 1 - Data loaded: 29.93
23:33:11.785 -> Id: 1 - Type: 3 - Data loaded: 68.00
23:33:11.785 -> Id: 1 - Type: 4 - Data loaded: 1013.34
23:33:11.785 -> Id: 1 - Type: 23 - Data loaded: -0.75
23:33:12.056 -> Id: 1 - Type: 34 - Data loaded: 8.16
23:33:12.356 -> Sending FDRS Packet!
23:33:12.356 -> Transmitting LoRa message of size 41 bytes with CRC 0xa9b5 to gateway 0xee02. Retries remaining: 2, Ack Ok 100.00%
23:33:12.841 ->  failed, code -5

**Reply by javrolijk** [reply]
@timmbogner 
I tested the example provided. I've uploaded the transmit on a TTGO Lora ESP32 and the receive code snippet in the Cubecell. The screenshot below shows that the Cubecell receives the packet as it should.

![image](https://github.com/timmbogner/Farm-Data-Relay-System/assets/70583691/50041715-1b38-45dc-a78b-f21ea2c2ac00)

**Reply by timmbogner** [reply]
@javrolijk What about the opposite, does anything fail when sending from the CubeCell?

Also, I just updated the code for those testing examples. It was a fix I made months ago but forgot to commit, so let me know if it broke anything.

**Reply by javrolijk** [reply]
@timmbogner The Cubecell sends data through Lora Ok. Tested the code for my application and all works. Only when enabling '#define LORA_ACK' it will give the -5 fault code after 3-4 cycles. Uploading the transmit example code on the Cubecell and the receive example code on the TTGO ESP32 work flawlessly, with no error code.

**Reply by timmbogner** [reply]
I think my best guess is that there is some glitch within the timing of everything. I have a general idea of where something could be going wrong, but I don't have the time or hardware to sort it out. I don't know how it all works, but is there an interrupt handler for the GPS NMEA data that might interrupt FDRS? Can you somehow fully disable the GPS and try it?

I would check out what boards LilyGo has with integrated GPS and LoRa. Unlike Heltec, LilyGo has been happy to send me samples of their products in order to help me develop for them 😏. I don't have anything from the T-Beam line, but that is what I would look at.

**Reply by javrolijk** [reply]
@timmbogner sounds like a plan.
If you want and have the time. My code package is in a zip file, and you can take a look. Perhaps you can find something that I may not be explaining correctly.
